src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -0,0 +1,774 @@
+                ");
+            }
+
+            return this.GetSqlCommandWithParameters(renewLeasesQuery.ToString(), connection, null);
+        }
+
+        /// <summary>
+        /// Builds the query to release leases on the rows in "_rows" after successful invocation of the user's function
+        /// (<see cref="RunChangeConsumptionLoopAsync()"/>).
+        /// </summary>
+        /// <param name="connection">The connection to add to the returned SqlCommand</param>
+        /// <param name="transaction">The transaction to add to the returned SqlCommand</param>
+        /// <returns>The SqlCommand populated with the query and appropriate parameters</returns>
+        private SqlCommand BuildReleaseLeasesCommand(SqlConnection connection, SqlTransaction transaction)
+        {
+            var releaseLeasesQuery = new StringBuilder("DECLARE @current_change_version bigint;\n");
+
+            for (int index = 0; index < this._rows.Count; index++)
+            {
+                string changeVersion = this._rows[index]["SYS_CHANGE_VERSION"];
+
+                releaseLeasesQuery.Append($@"
+                    SELECT @current_change_version = ChangeVersion
+                    FROM {this._workerTableName} WITH (TABLOCKX)
+                    WHERE {this._rowMatchConditions[index]};
+
+                    IF @current_change_version <= {changeVersion}
+                        UPDATE {this._workerTableName} WITH (TABLOCKX) 
+                        SET ChangeVersion = {changeVersion}, AttemptCount = 0, LeaseExpirationTime = NULL
+                        WHERE {this._rowMatchConditions[index]};
+                ");
+            }
+
+            return this.GetSqlCommandWithParameters(releaseLeasesQuery.ToString(), connection, transaction);
+        }
+
+        /// <summary>
+        /// Builds the command to update the global version number in _globalStateTable after successful invocation of
+        /// the user's function. If the global version number is updated, also cleans the worker table and removes all
+        /// rows for which ChangeVersion <= newLastSyncVersion.
+        /// </summary>
+        /// <param name="connection">The connection to add to the returned SqlCommand</param>
+        /// <param name="transaction">The transaction to add to the returned SqlCommand</param>
+        /// <param name="newLastSyncVersion">The new LastSyncVersion to store in the _globalStateTable for this _userTableName</param>
+        /// <returns>The SqlCommand populated with the query and appropriate parameters</returns>
+        private SqlCommand BuildUpdateTablesPostInvocation(SqlConnection connection, SqlTransaction transaction, long newLastSyncVersion)
+        {
+            string workerTableJoinCondition = string.Join(" AND ", this._primaryKeyColumns.Select(col => $"c.{col} = w.{col}"));
+
+            // TODO: Need to think through all cases to ensure the query below is correct, especially with use of < vs <=.
```

Comment:
Has this been done?

Reply:
Still pending. The comment must be added by me when I did not know about the design document. I will be able to check this tomorrow.

Reply:
The usage of `<=` is correct. Changes whose versions equal `newLastSyncVersion` should also be counted as unprocessed. I will remove the comment. Though I still need to take some time to wrap my head completely around why the last-sync version computation is the way it is.

Reply:
Sounds good, that's definitely something that we need to spend more time going over in detail too.